### ISSUE TITLE
Deleting last file in the root of a bucket creates and leaves forever an empty folder placeholder.

### ISSUE THREAD

**Issue by GaryAustin1** [supabase/supabase#9155]
# Bug report

## Describe the bug

If a bucket has one file and you delete it thru the Dashboard UI a fake empty folder placeholder is generated.
![image](https://user-images.githubusercontent.com/54564956/192173998-ac8be620-8165-47d6-bba0-73ab93eb037a.png)

If a new file is added, the fake placeholder stays around.  This is different than adding a file to a non-root folder in that the place holder for those goes away.  In the image below you can see the cats folder which was an empty place holder is replaced with just the filepath.
![image](https://user-images.githubusercontent.com/54564956/192174468-fbc38205-6801-4cae-97c1-7c21d7ddfedd.png)

There is also no way to delete this easily without going to the SQL editor as it does not show up in the storage UI and you can't edit storage tables in the Table UI.  Using the SQL editor on the storage objects table is risky.

A user in Discord was having issues with this and was trying to count the files coming back from list, but this extra entry is also returned.  https://discord.com/channels/839993398554656828/1023730313865724005

Note that deleting the last file from the API does not create this fake root folder placeholder.

## To Reproduce

Do the above steps.

## Expected behavior

Constant behavior with sub-folders and probably a way to list with out returning folder placeholders.

**Comment by inian** [comment]
Moving this to supabase repo since this is a frontend issue

**Comment by GaryAustin1** [comment]
@inian I did not see an issue in storage-api, is there anything in the works to have a list files option without showing folders (empty or otherwise)?

**Comment by inian** [comment]
Hey Gary, the .emptyFolderPlaceholder is just created by the dashboard, storage-api has no awareness of it - it is just another file. We do this to mimic folder support since s3 doesn't have the concept of folders. The client (dashboard) is supposed to filter out this file and not display it, but looks like there are some regressions there.

**Comment by GaryAustin1** [comment]
Yeah, that is always a fun discussion in Discord.  I deleted my second comment on list() having option to ignore "sub folders" because of that reason.  Wanted to think about it more from what users are doing with it.  
But forgot you still get the email.  Thanks

**Comment by GaryAustin1** [comment]
@inian sorry for another post.  s3 does not have a concept of this true.  But storage-api does have concept of this because the empty placeholders show up in list() as file objects.  This case is a bit special as the root placeholder stays around even with files added.

This root case can be fixed, but the empty sub folder place holders would be a big impact on the UI as they allow creating them outright.  

Not sure if some option on .list() is not appropriate to filter them.

**Comment by joshenlim** [comment]
> The client (dashboard) is supposed to filter out this file and not display it, but looks like there are some regressions there 
Yup the dashboard does filter out that particular file, the screenshot from the original post is from the table editor of the storage schema

@GaryAustin1 will the right fix (from the UI side) then be only having the `.emptyFolderPlaceholder` if the folder is empty and not specifically deleted? so we delete `.emptyFolderPlaceholder` if the user uploads a file to an empty folder, and add it back if the folder is emptied?

**Comment by GaryAustin1** [comment]
I would think not generating it all in the root is the right thing.  The storage UI never shows it in the root like it does for subfolders so seems useless and is confusing if you use API list() it shows up as a file.

**Comment by GaryAustin1** [comment]
closing this as I don't think it will be fixed, but hopefully searching will find this explanation.

**Comment by sagar-jani** [comment]
I faced a similar issue while displaying the files from bucket.
![image](https://github.com/supabase/supabase/assets/10905009/23be4238-66ba-4cc0-ad84-28781918ef37)

**Comment by RaoulRSDSchipper** [comment]
I found the same problem. When deleting everything in the storage bucket with the website, it generates a .emptyFolderPlaceholder that will be returned by the api when using .list. When everything in the bucket is deleted with the api and one then uses .list on the bucket, no placeholder is returned.

**Comment by encima** [comment]
Re-opening as users have reported seeing this again

**Comment by Hallidayo** [comment]
Hi everyone - just checking if you are still experiencing this issue? I've tried to recreate on a couple of project but the issue seems okay now?

**Comment by vincentjflorio** [comment]
I am not seeing it in the dashboard but it is being returned sometimes on a fetch (if that's an out of scope comment to make lemme know and I'll delete this, or simply ignore it)

I use list with a limit of 1 to retrieve a single value, because the only non-array retrieval I am away of is to force a download, and we just want to see what object is 'on file' for a given user in a given bucket.

Because the JS client permits me to do that and doesn't know the character of what that first item is, it will sometimes see the placeholder, and select that, and set that to be the value for the upload field :)

**Comment by Andreasams** [comment]
This is still a thing. 

I have a function creating a set of folders in my storage when a new row is added to a table. 
When adding images to one of the folders, the information about those images are then added to a gallery table.
Whenever I delete the last image inside one of these folders, the name of the final image to be deleted is replaced with .emptyFolderPlaceholder. 

Using something like the code below to ignore it has no effect.

 IF new.name = '.emptyFolderPlaceholder' THEN
        RETURN new;
    END IF;
   
  I've also added a CHECK on my gallery table, but the result of doing this is that the entire folder is deleted because the .emptyFolderPlaceholder seems to lack a place to exist 
  
  ALTER TABLE public.gallery_image
ADD CONSTRAINT check_not_placeholder
CHECK (image_name != '.emptyFolderPlaceholder');

So at this stage, I either have to live with my table filling up with these entries so as to maintain the folder integrity inside my storage and not get issues on the frontend because a folder is missing - or I have to try to find other ways to prune my table. That's what it looks like right now.

**Comment by Andreasams** [comment]
When I have the folders created automatically, a "placeholder.json" file containing just {} and taking up 2 bytes are added to the folders so as to establish the folders. 

I have been able to make sure these are not added to my gallery table, so the solution for me is going to be to simply keep these placeholder files inside my image galleries and filter them out from queries such that I never delete the last file inside a folder. 

That will prevent my gallery table from being "polluted" with the emptyFolderPlaceholder entries while not requiring much storage space.

**Comment by donghoon-song** [comment]
How can I remove it manually? Even I uploaded an item, it still exist.

**Comment by jgturner** [comment]
Run this in the SQL Editor. May have to modify it but it got rid of it. 

delete from storage.objects
where bucket_id = 'images' 
and name like '%.emptyFolderPlaceholder';

**Comment by noook** [comment]
I managed to reproduce with the following steps:
- [studio] Created a folder with `auth.uid()`
- [studio] Uploaded 2 documents in it.
- [studio] Realized I forgot to create a subfolder, creating it
- [studio] Moving the 2 documents from `/${auth.uid()` to `${auth.uid()/documents`
- [supabase-js] `.list()` returns the empty placeholder in addition to the 2 documents.

**Comment by PrestigiousF** [comment]
Just ran across this issue. Similar way to the above reply. I moved documents from one folder to a newly created folder and I believe it was transferred from the old.

**Comment by TSCOfficial** [comment]
I got the same issue today. I created a new bucket and then added three files (via drag'n'drop on the dashboard)
THe .emptyFolderPlaceholder doesnt show up in the bucket view but in the object table and in .list() :,)

**Comment by awaseem** [comment]
I can't seem to reproduce this issue locally. I added files and created folders in a bucket, delete said files and when I fetched them in the API I just get back an empty array. Closing this as no reproducible. If someone has new steps, would love to take a look!

**Comment by julianlangweb** [comment]
still encountering this. and also where .emptyPlaceholderFolder exists the supabase-js retrieval of uploaded documents in this order seems broken.

**Comment by lastmjs** [comment]
This just happend to me

**Comment by nicholas-masini** [comment]
Happened to me as well
